doc/man3/EVP_PKEY_asn1_get_count.pod

```diff
@@ -0,0 +1,72 @@
+=pod
+
+=head1 NAME
+
+EVP_PKEY_asn1_find,
+EVP_PKEY_asn1_find_str,
+EVP_PKEY_asn1_get_count,
+EVP_PKEY_asn1_get0,
+EVP_PKEY_asn1_get0_info
+- enumerate public key ASN.1 methods
+
+=head1 SYNOPSIS
+
+ #include <openssl/evp.h>
+
+ int EVP_PKEY_asn1_get_count(void);
+ const EVP_PKEY_ASN1_METHOD *EVP_PKEY_asn1_get0(int idx);
+ const EVP_PKEY_ASN1_METHOD *EVP_PKEY_asn1_find(ENGINE **pe, int type);
+ const EVP_PKEY_ASN1_METHOD *EVP_PKEY_asn1_find_str(ENGINE **pe,
+                                                    const char *str, int len);
+ int EVP_PKEY_asn1_get0_info(int *ppkey_id, int *pkey_base_id,
+                             int *ppkey_flags, const char **pinfo,
+                             const char **ppem_str,
+                             const EVP_PKEY_ASN1_METHOD *ameth);
+
+=head1 DESCRIPTION
+
+EVP_PKEY_asn1_count() returns a count of the number of public key
+ASN.1 methods available: it includes standard methods and any methods
+added by the application.
+
+EVP_PKEY_asn1_get0() returns the public key ASN.1 method B<idx>.
+The value of B<idx> must be between zero and EVP_PKEY_asn1_get_count()
+- 1.
+
+EVP_PKEY_asn1_find() looks up the B<EVP_PKEY_ASN1_METHOD> with NID
+B<type>.
+It searches through the methods added by the application first, and
+then through the standard methods.
+
+EVP_PKEY_asn1_find_str() looks up the B<EVP_PKEY_ASN1_METHOD> with PEM
+type string B<str>.
+Just like EVP_PKEY_asn1_find(), it searches through the methods added
+by the application first, and then through the standard methods.
```

Comment:
really?

Reply:
[Have a look for yourself](https://github.com/openssl/openssl/blob/master/crypto/asn1/ameth_lib.c#L55-L70) ;-)

Reply:
to me this looks like the standard methods win:

    for (i = 0; i < EVP_PKEY_asn1_get_count(); i++) {
        ameth = EVP_PKEY_asn1_get0(i);
        if (ameth->pkey_flags & ASN1_PKEY_ALIAS)
            continue;
        if (((int)strlen(ameth->pem_str) == len)
            && (strncasecmp(ameth->pem_str, str, len) == 0))

Reply:
FYI: The comment refers to EVP_PKEY_asn1_find_str

Reply:
Ah, you are quite correct!  That's...  not very consistent.  I guess the idea was that everyone would play nice with each other and not try to compete over the same algorithms, but rather that application would handle their own *different* algo.

When I look at `EVP_PKEY_asn1_add0` and the way it sorts the stack every time a new method is added, it seems clear to me that the original line of thought is one of no competition, and to make sure that is always true, `EVP_PKEY_asn1_add0` should refuse to add a method for an algo that already has one.  In that case, the exact order in which these methods are found don't matter at all.

Reply:
Yes, that would be a significant improvement.

Reply:
Should the check in the add method also look for duplicate strings?